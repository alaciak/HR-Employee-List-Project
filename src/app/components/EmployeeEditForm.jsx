import React from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';

class EmployeeEditForm extends React.Component {
  static propTypes = {
    history: PropTypes.object.isRequired
  }

  constructor(props) {
    super(props);
    this.state = {
      firstname:'',
      lastname:'',
      position: '',
      role: '',
      experience: '',
      shortdescript: '',
      longdescript:'',
      loading: true
    }
  }

  componentDidMount() {
    fetch(`http://localhost:3000/employees/${this.props.match.params.id}`).then(r => r.json()).then(data => {
      this.setState({
        id: data.id,
        firstname: data.firstname,
        lastname: data.lastname,
        position: data.position,
        role: data.role,
        experience: data.experience,
        shortdescript: data.shortdescript,
        longdescript: data.longdescript,
        loading: false
      });
    }).catch(function(error) {
      console.log(error);
    });
  }

  handleOnChange = e => {
    const target = e.target;
    const value = target.value;
    const name = target.name;
    this.setState({
      [name]: value,
    });
  }

  handleOnClickCancel  = e => {
    e.preventDefault();
    this.props.history.push('/');
  }

  handleonClickUpdate = e => {
    e.preventDefault();
    const updatedEmployee = {
      id: this.state.id,
      firstname: this.state.firstname,
      lastname: this.state.lastname,
      position: this.state.position,
      role: this.state.role,
      experience: this.state.experience,
      shortdescript: this.state.shortdescript,
      longdescript: this.state.longdescript,
    }
    fetch(`http://localhost:3000/employees/${this.props.match.params.id}`, {
      method: 'PUT',
      body: JSON.stringify(updatedEmployee),
      headers: {
        "Content-Type": "application/json"
      },
    }).then(data => {
      this.props.history.push('/');
    }).catch(function(error) {
      console.log(error);
    });
  }

  render() {
    return <form action='' method='POST'>
      <label>First name:
        <input type='text' value={ this.state.firstname } placeholder={ this.state.firstname } disabled ></input>
      </label>
      <label>Last name:
        <input type='text' value={ this.state.lastname } placeholder={ this.state.lastname } disabled ></input>
      </label>
      <label>Role:
        <input type='text' value={ this.state.role } placeholder={ this.state.role }  disabled ></input>
      </label>
      <label>Position:
        <input type='text' value={ this.state.position } placeholder={ this.state.position } disabled ></input>
      </label>
      <label>Experience:
        <input type='text' value={ this.state.experience } placeholder={ this.state.experience } name='experience' onChange={ this.handleOnChange } ></input>
      </label>
      <label>Short description:
        <input type='text' value={ this.state.shortdescript } placeholder={ this.state.shortdescript } name='shortdescript' onChange={ this.handleOnChange } ></input>
      </label>
      <label>Long description:
        <input type='text' value={ this.state.longdescript } placeholder={ this.state.longdescript } name='longdescript' onChange={ this.handleOnChange } ></input>
      </label>
      <button type='button' onClick={ this.handleonClickUpdate }>SAVE</button>
      <button type='button' onClick={ this.handleOnClickCancel }>CANCEL</button>
    </form>
  }
}

module.exports = EmployeeEditForm;
