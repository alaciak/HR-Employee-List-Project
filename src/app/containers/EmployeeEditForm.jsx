import React from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { getEmployee, updateEmployee } from '../actions/employeeEditFormActions';

export class EmployeeEditForm extends React.Component {
  static propTypes = {
    history: PropTypes.object.isRequired
  }

  constructor(props) {
    super(props);
    this.state = {
      employee: {
        firstname: '',
        lastname: '',
        position: '',
        role: '',
        experience: '',
        shortdescript: '',
        longdescript: '',
      },
      formAlertDisplay: 'none',
      formExperienceAlertDisplay: 'none'
    }
  }

  componentDidMount() {
    this.props.getEmployee(this.props.match.params.id);
    if(this.props.userType === 'admin') {
      this.setState({
        disabled: false
      });
    } else {
      this.setState({
        disabled: true
      });
    }
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      employee: nextProps.employee
    });
    if(nextProps.userType === 'admin') {
      this.setState({
        disabled: false
      });
    } else {
      this.setState({
        disabled: true
      });
    }
  }

  handleOnChange = e => {
    const value = e.target.value;
    const name = e.target.name;
    this.setState({
      employee: {
        ...this.state.employee,
        [name]: value
      }
    });
  }

  handleOnClickCancel = e => {
    e.preventDefault();
    this.props.history.push('/');
  }

  handleOnClickUpdate = e => {
    console.log(typeof Number(this.state.employee.experience));
    e.preventDefault();
    if(!this.state.employee.role || !this.state.employee.position || !this.state.employee.experience || !this.state.employee.shortdescript || !this.state.employee.longdescript) {
      this.setState({
        formAlertDisplay: 'block'
      });
    } else if(typeof Number(this.state.employee.experience) !== 'number') {
      this.setState({
        formExperienceAlertDisplay: 'block'
      });
    } else {
      this.props.updateEmployee(this.state.employee, this.props.history);
    }
  }

  render() {
    return (
      <section className='employee-data-form container'>
        <div className='row'>
          <form action='' name='employee-data' className='employee-data-form_edit col-12'>
            <fieldset>
              <legend>Employee Data</legend>
              <label className='employee-data-form_edit-label'>First name:
                <input type='text' value={ this.state.employee.firstname } placeholder={ this.state.employee.firstname } disabled></input>
              </label>
              <label className='employee-data-form_edit-label'>Last name:
                <input type='text' value={ this.state.employee.lastname } placeholder={ this.state.employee.lastname } disabled></input>
              </label>
              <label className='employee-data-form_edit-label'>Role:
                <input type='text' value={ this.state.employee.role } placeholder={ this.state.employee.role } disabled={ this.state.disabled } name='role' onChange={ this.handleOnChange }></input>
              </label>
              <label className='employee-data-form_edit-label'>Position:
                <input type='text' value={ this.state.employee.position } placeholder={ this.state.employee.position } disabled={ this.state.disabled } name='position' onChange={ this.handleOnChange }></input>
              </label>
              <label className='employee-data-form_edit-label'>Experience (months):
                <input type='text' value={ this.state.employee.experience } placeholder={ this.state.employee.experience}  name='experience' onChange={ this.handleOnChange }></input>
              </label>
              <label className='employee-data-form_edit-label'>Short description:
                <textarea className='employee-data-form_edit-description' type='text' value={ this.state.employee.shortdescript } placeholder={ this.state.employee.shortdescript } name='shortdescript' onChange={this.handleOnChange}></textarea>
              </label>
              <label className='employee-data-form_edit-label'>Long description:
                <textarea className='employee-data-form_edit-description' type='text' value={ this.state.employee.longdescript } placeholder={ this.state.employee.longdescript } name='longdescript' onChange={this.handleOnChange}></textarea>
              </label>
            </fieldset>
          </form>
          <div className='row employee-data-form_buttons'>
            <button className='btn btn-save' type='button' onClick={ this.handleOnClickUpdate }>SAVE</button>
            <button className='btn btn-cancel' type='button' onClick={ this.handleOnClickCancel }>CANCEL</button>
          </div>
        </div>
        <div className='row employee-data-form_alert-message' style={{ display: this.state.formAlertDisplay }}>All the fields must be completed before saving the form</div>
        <div className='row employee-data-form_alert-message' style={{ display: this.state.formExperienceAlertDisplay }}>Please provide number of months as an experience</div>
      </section>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    employee: state.employeeEditFormReducer.employee,
    userType: state.userTypeSelectionReducer.userType
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    getEmployee: (employeeId) => {
      dispatch(getEmployee(employeeId));
    },
    updateEmployee: (employee, history) => {
      dispatch(updateEmployee(employee, history));
    }
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(EmployeeEditForm));
