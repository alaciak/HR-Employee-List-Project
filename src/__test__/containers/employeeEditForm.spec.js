import React from 'react';
import { EmployeeEditForm } from '../../app/containers/EmployeeEditForm.jsx';
import { shallow, render } from 'enzyme';
import toJson from 'enzyme-to-json';
import createRouterContext from 'react-router-test-context';

describe('EmployeeEditForm', () => {

  it('should render if the data are fetched from the server', () => {

    const testEmployee = {
      id: 1,
      firstname: 'testName',
      lastname: 'testLastname',
      position: 'testPosition',
      role: 'testRole',
      experience: 'testExperience',
      shortdescript: 'testShortdescript',
      longdescript: 'testLongDescript'
    };

    const context = createRouterContext();
    EmployeeEditForm.propTypes = {
      router: React.PropTypes.object
    };

    const wrapper = render(
        <EmployeeEditForm loading={ false } employee={ testEmployee }/>,
        { context }
    );

    expect(toJson(wrapper)).toMatchSnapshot();

  });

});
