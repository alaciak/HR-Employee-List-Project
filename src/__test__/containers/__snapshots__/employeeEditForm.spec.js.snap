// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmployeeEditForm should render if the data are fetched from the server 1`] = `
<section
  class="employee-data-form container"
>
  <div
    class="row"
  >
    <form
      action=""
      class="employee-data-form_edit col-12"
      name="employee-data"
    >
      <fieldset>
        <legend>
          Employee Data
        </legend>
        <label
          class="employee-data-form_edit-label"
        >
          First name:
          <input
            disabled=""
            placeholder=""
            type="text"
            value=""
          />
        </label>
        <label
          class="employee-data-form_edit-label"
        >
          Last name:
          <input
            disabled=""
            placeholder=""
            type="text"
            value=""
          />
        </label>
        <label
          class="employee-data-form_edit-label"
        >
          Role:
          <input
            name="role"
            placeholder=""
            type="text"
            value=""
          />
        </label>
        <label
          class="employee-data-form_edit-label"
        >
          Position:
          <input
            name="position"
            placeholder=""
            type="text"
            value=""
          />
        </label>
        <label
          class="employee-data-form_edit-label"
        >
          Experience (months):
          <input
            name="experience"
            placeholder=""
            type="text"
            value=""
          />
        </label>
        <label
          class="employee-data-form_edit-label"
        >
          Short description:
          <textarea
            class="employee-data-form_edit-description"
            name="shortdescript"
            placeholder=""
            type="text"
          />
        </label>
        <label
          class="employee-data-form_edit-label"
        >
          Long description:
          <textarea
            class="employee-data-form_edit-description"
            name="longdescript"
            placeholder=""
            type="text"
          />
        </label>
      </fieldset>
    </form>
    <div
      class="row employee-data-form_buttons"
    >
      <button
        class="btn btn-save"
        type="button"
      >
        SAVE
      </button>
      <button
        class="btn btn-cancel"
        type="button"
      >
        CANCEL
      </button>
    </div>
  </div>
</section>
`;
